### PR TITLE
Add kernel that captures an array and length

### PR DESCRIPTION
**Pull request by MrBurmark**
# Add kernel that captures an array and uses it with a runtime length

Add a kernel that captures an array and only uses a runtime sized portion of that array.
This pattern doesn't optimize well with RAJA Cuda.  I have tried this with nvcc 10.1 and 11.7 and with both the the RAJA_CUDA variant performs about 3.3x worse than the Base_CUDA and Lambda_CUDA variants. The Base_CUDA and Lambda_CUDA variants perform as expected for a bandwidth kernel and achieve a bit more bandwidth than DAXPY. Looking at the assembly the RAJA variant of the kernel is copying the array to local memory while the base and lambda variants are not performing the extra copy.
This pattern performs the same for all HIP variants. That performance is below expectations for a bandwidth bound kernel, about 0.8x the bandwidth of DAXPY. Both using pragma unroll 4 and using the compile time length of the array didn't significantly improve performance. Using a local variable for accumulation improved performance to within expectations.

I ran this via a command like `./bin/raja-perf.exe -k DAXPY ARRAY_OF_PTRS`

- This PR is a feature
- It does the following (modify list as needed):
  - Adds example of partial array usage at the request of myself

### CODE REVIEW

**Comment by rhornung67** [comment]
@MrBurmark I merged in develop and fixed merge conflicts in this. Will it be ready to go when checks pass?

**Comment by MrBurmark** [comment]
Nope, it needs to be updated in a couple ways.

**Comment by MrBurmark** [comment]
@rhornung67 I got this fixed, take a look and we can merge it.

File: src/basic-kokkos/ARRAY_OF_PTRS-Kokkos.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
```
Comment (rhornung67) [inline src/basic-kokkos/ARRAY_OF_PTRS-Kokkos.cpp:2]:
```suggestion
// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
```

File: src/basic/ARRAY_OF_PTRS-Cuda.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
```
Comment (rhornung67) [inline src/basic/ARRAY_OF_PTRS-Cuda.cpp:2]:
```suggestion
// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
```

File: src/basic/ARRAY_OF_PTRS-OMPTarget.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
```
Comment (rhornung67) [inline src/basic/ARRAY_OF_PTRS-OMPTarget.cpp:2]:
```suggestion
// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
```

File: src/basic/ARRAY_OF_PTRS-Seq.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
```
Comment (rhornung67) [inline src/basic/ARRAY_OF_PTRS-Seq.cpp:2]:
```suggestion
// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
```

File: src/basic/ARRAY_OF_PTRS.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
```
Comment (rhornung67) [inline src/basic/ARRAY_OF_PTRS.cpp:2]:
```suggestion
// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
```

File: src/basic/ARRAY_OF_PTRS.hpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
```
Comment (rhornung67) [inline src/basic/ARRAY_OF_PTRS.hpp:2]:
```suggestion
// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
```

**Review by rhornung67 — COMMENTED** [review]
Only a few cosmetic changes to fix the file header copyright year.

**Comment by MrBurmark** [comment]
@rhornung67 I applied your suggestions.

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**